### PR TITLE
Added user toggle to schedulers.

### PR DESCRIPTION
**Pull request by gtmacdonald**
Hello!

I have added the ability for the user to toggle a scheduler.

In the schedulers UI there's now a new column with checkboxes to enable or disable a scheduler. Periodic schedulers will print a log message if a build is being skipped. Force schedulers will have their build button disabled and display a red warning banner reminding the user the scheduler is disabled.

To do this I've added an enabled column to the schedulers table in the database. I've added unit tests by mirroring what was done for the step.hidden attribute which was similarly added. I couldn't see where to add docs. Let me know if you need me to add some.

force_sched

scheduler

Please let me know if you need anything. :) Thx!

### CODE REVIEW

File: master/buildbot/data/forceschedulers.py [hunk 1]
```diff
@@ -27,7 +27,8 @@ def forceScheduler2Data(sched):
                name=unicode(sched.name),
                button_name=unicode(sched.buttonName),
                label=unicode(sched.label),
-               builder_names=map(unicode, sched.builderNames))
+               builder_names=map(unicode, sched.builderNames),
```
Comment (sa2ajj) [inline master/buildbot/data/forceschedulers.py]: while you are at it, do you think you could convert this `map` to a list comprehension?  (that'd be a bit more readable)

File: master/buildbot/schedulers/base.py [hunk 2]
```diff
@@ -86,12 +86,29 @@ def __init__(self, name, builderNames, properties=None,
...
+    # activity handling
+    @defer.inlineCallbacks
     def activate(self):
-        return defer.succeed(None)
+        # even if we aren't called via _activityPoll(), at this point we
+        # need to ensure the service id is set correctly
+        if self.serviceid is None:
+            self.serviceid = yield self._getServiceId()
+            assert self.serviceid is not None
+
```
Comment (sa2ajj) [inline master/buildbot/schedulers/base.py]: there should only be one empty line for separating logical blocks.

File: master/buildbot/schedulers/base.py [hunk 3]
```diff
@@ -150,6 +177,13 @@ def _changeCallback(self, key, msg, fileIsImportant, change_filter,
                     'not processed by scheduler %(name)s',
                     codebase=change.codebase, name=self.name)
             return
+
+        if not self.enabled:
+            log.msg(format='ignoring change from %(name)s because ' \
```
Comment (sa2ajj) [inline master/buildbot/schedulers/base.py]: no need for `\` at the end of the line

File: master/buildbot/schedulers/timed.py [hunk 4]
```diff
@@ -130,6 +130,12 @@ def gotChange(self, change, important):
 
     @defer.inlineCallbacks
     def startBuild(self):
+        if not self.enabled:
+            log.msg(format='ignoring build from %(name)s because scheduler' \
```
Comment (sa2ajj) [inline master/buildbot/schedulers/timed.py]: ditto

File: master/buildbot/schedulers/timed.py [hunk 5]
```diff
@@ -407,6 +413,12 @@ def trigger(self, waited_for, sourcestamps=None, set_props=None,
 
     @defer.inlineCallbacks
     def startBuild(self):
+        if not self.enabled:
+            log.msg(format='ignoring build from %(name)s because scheduler ' \
+                           'has been disabled by the user',
```
Comment (sa2ajj) [inline master/buildbot/schedulers/timed.py]: ditto

**Comment by sa2ajj** [comment]
LGTM otherwise

(I wish GitHub would allow just one review "item" with all messages inside)

**Comment by gtmacdonald** [comment]
Great! I'll make those changes and resubmit.

**Comment by gtmacdonald** [comment]
Aw so close. Looks like postgres wants a tiny int instead of a boolean for some reason.

**Comment by gtmacdonald** [comment]
Coverage doesn't look so good but I think it's being pulled down by the worker code, which isn't related AFAIK.

File: .travis.yml [hunk 6]
```diff
@@ -126,7 +126,7 @@ install:
 
   # Note pylint version is pinned because newer versions can't import zope.interface - http://www.logilab.org/92792
   # astroid > 1.4 is incompatible with pylint 1.1.0
-  - "[ $TESTS != lint ] || pip install pylint==1.1.0 pep8==1.5.7 flake8  astroid==1.3.8"
+  - "[ $TESTS != lint ] || pip install pylint==1.1.0 pep8==1.5.7 flake8==2.5.5  astroid==1.3.8"
```
Comment (tardyp) [inline .travis.yml]: this will clash with https://github.com/buildbot/buildbot/pull/2280

**Comment by tardyp** [comment]
I like the idea of dynamic enable/disable of stuff.

I think this would make sense also for workers. To disable them instead of disconnecting (will allow to implement autorestart on the worker side)


I am a bit annoyed to merge this big change before rc1 though.
Would you care if I merge it for 0.9.1?

**Comment by gtmacdonald** [comment]
Hi Pierre! I'm glad it'll be useful for other things. I wasn't sure if it fit within the bigger picture or not. I don't mind waiting. :) Thx!

**Comment by tardyp** [comment]
rc1 has been released for long, and I forgot this one.
Please rebase, and I will merge it.

**Comment by gtmacdonald** [comment]
Okie dokie.

**Comment by gtmacdonald** [comment]
Finished rebase.

File: master/buildbot/data/schedulers.py [hunk 7]
```diff
@@ -55,6 +56,14 @@ def get(self, resultSpec, kwargs):
         defer.returnValue((yield self.db2data(dbdict))
                           if dbdict else None)
 
+    @defer.inlineCallbacks
+    def control(self, action, args, kwargs):
+        if action == 'setenabled':
+            schedulerid = kwargs['schedulerid']
+            enabledValue = args['enabledValue']
+
+            yield self.master.data.updates.schedulerSetEnabled(schedulerid, enabledValue)
```
Comment (tardyp) [inline master/buildbot/data/schedulers.py]: this requires a unit test

File: master/buildbot/db/schedulers.py [hunk 8]
```diff
@@ -28,6 +28,13 @@ class SchedulerAlreadyClaimedError(Exception):
 class SchedulersConnectorComponent(base.DBConnectorComponent):
     # Documentation is in developer/db.rst
 
+    def setEnabled(self, schedulerid, enabled_value):
+        def thd(conn):
+            tbl = self.db.model.schedulers
+            q = tbl.update(whereclause=(tbl.c.id == schedulerid))
+            conn.execute(q, enabled=int(enabled_value))
+        return self.db.pool.do(thd)
```
Comment (tardyp) [inline master/buildbot/db/schedulers.py]: unit test please

File: master/buildbot/data/schedulers.py [hunk 9]
```diff
@@ -81,13 +90,29 @@ class Scheduler(base.ResourceType):
...
+    def generateEvent(self, schedulerid, event):
+        scheduler = yield self.master.data.get(('schedulers', str(schedulerid)))
+        self.produceEvent(scheduler, event)
+
+    @base.updateMethod
+    @defer.inlineCallbacks
+    def schedulerSetEnabled(self, schedulerid, enabled_value):
+        yield self.master.db.schedulers.setEnabled(
+            schedulerid=schedulerid, enabled_value=enabled_value)
+        yield self.generateEvent(schedulerid, 'updated')
```
Comment (tardyp) [inline master/buildbot/data/schedulers.py]: unit tests

**Comment by tardyp** [comment]
please add some unit tests for the added python code, and also a line in the release note is required for such change

**Comment by tardyp** [comment]
@gtmacdonald do you intend to finish this PR?

**Comment by gtmacdonald** [comment]
I do, we'll need this feature. I've just been swamped with school and work. And it hasn't been a priority because we're only partially converted over to buildbot. Sorry for the delay.

**Comment by gtmacdonald** [comment]
@tardyp Hi Pierre, Sorry for the delay. I've rebased with unit tests and release notes. Please let me know if there's anything else to do.

**Comment by gtmacdonald** [comment]
@tardyp I've fixed the extra space lint issue. It doesn't seem to be rebuilding for some reason. There's also something about an unclean reactor and workers that I'm not sure is relevant.

File: master/buildbot/test/unit/test_db_schedulers.py [hunk 10]
```diff
@@ -44,11 +44,53 @@ class Tests(interfaces.InterfaceTests):
...
+            name='schname',
+            enabled=True,
+            masterid=13))
+
     def test_signature_classifyChanges(self):
         @self.assertArgSpecMatches(self.db.schedulers.classifyChanges)
         def classifyChanges(self, schedulerid, classifications):
             pass
 
+    # def setEnabled(self, schedulerid, enabled_value):
```
Comment (tardyp) [inline master/buildbot/test/unit/test_db_schedulers.py]: I think this is unintendended left overs

File: master/buildbot/data/schedulers.py [hunk 11]
```diff
@@ -81,13 +90,30 @@ class Scheduler(base.ResourceType):
...
     entityType = EntityType(name)
 
+    @defer.inlineCallbacks
+    def generateEvent(self, schedulerid, event):
+        scheduler = yield self.master.data.get(('schedulers', str(schedulerid)))
+        self.produceEvent(scheduler, event)
+
+    @base.updateMethod
+    @defer.inlineCallbacks
+    def schedulerSetEnabled(self, schedulerid, enabled_value):
```
Comment (tardyp) [inline master/buildbot/data/schedulers.py]: I would rather call this API 

    def schedulerEnable(self, schedulerid, enable=True)

File: master/buildbot/data/schedulers.py [hunk 12]
```diff
@@ -55,6 +56,14 @@ def get(self, resultSpec, kwargs):
         defer.returnValue((yield self.db2data(dbdict))
                           if dbdict else None)
 
+    @defer.inlineCallbacks
+    def control(self, action, args, kwargs):
+        if action == 'setenabled':
```
Comment (tardyp) [inline master/buildbot/data/schedulers.py]: I am not very fond of the setenabled name.

I think this would be better to change 'setenabled' -> 'enable' and 'enabledValue' to 'enable'

**Comment by gtmacdonald** [comment]
I finally got around to finishing this. Please merge. Thx!

File: master/buildbot/schedulers/base.py [hunk 13]
```diff
@@ -87,12 +87,29 @@ def __init__(self, name, builderNames, properties=None,
...
+        if self.serviceid is None:
+            self.serviceid = yield self._getServiceId()
+            assert self.serviceid is not None
+
+        schedulerData = yield self._getScheduler(self.serviceid)
+
+        if schedulerData:
+            self.enabled = schedulerData['enabled']
+
+        yield self.startConsumingEnabledChanges()
```
Comment (tardyp) [inline master/buildbot/schedulers/base.py]: changes is a thing in builbot its about changes to SCM. here we are watching events.

File: master/buildbot/schedulers/base.py [hunk 14]
```diff
@@ -151,6 +178,13 @@ def _changeCallback(self, key, msg, fileIsImportant, change_filter,
                     'not processed by scheduler %(name)s',
                     codebase=change.codebase, name=self.name)
             return
+
+        if not self.enabled:
```
Comment (tardyp) [inline master/buildbot/schedulers/base.py]: Shouldn't we rather stop the scheduler?
When it is restarted again it will consume changes that were left rather than just forget them

File: www/base/src/app/schedulers/schedulers.controller.coffee [hunk 15]
```diff
@@ -2,3 +2,23 @@ class schedulers extends Controller
...
+        # Assumes an order.
+        $scope.schedulers.onNew = (s) ->
+            $scope.enabledValues.push(s.enabled)
+
+        $scope.change = ->
+            i = 0
+            while i<$scope.enabledValues.length
+                oldValue = $scope.enabledValues[i]
+                scheduler = $scope.schedulers[i]
+                newValue = scheduler.enabled
```
Comment (tardyp) [inline www/base/src/app/schedulers/schedulers.controller.coffee]: looks too complicated.
why dont you pass the scheduler id in the change callback?

File: www/base/src/app/schedulers/schedulers.tpl.jade [hunk 16]
```diff
@@ -3,16 +3,22 @@
       uib-tab(heading="All schedulers")
         table.table.table-hover.table-striped.table-condensed
           tr
+              td Enabled
               td Scheduler Name
               td Master
           tr(ng-repeat='scheduler in schedulers | orderBy: ["name"]')
+              td
+                input(type="checkbox", ng-model="scheduler.enabled", ng-change="change()")
```
Comment (tardyp) [inline www/base/src/app/schedulers/schedulers.tpl.jade]: ng-change="change(scheduler)"

so that you get the schedulerid and the new value

File: master/buildbot/schedulers/base.py (hunk 14 above)
Comment (gtmacdonald) [inline master/buildbot/schedulers/base.py]: I'm not sure I understand. I'm following the same logic to ignore changes as the isFileImportant callback. So it will ignore the changes when the scheduler isn't on. Are you suggesting it SHOULD remember? If so then should I disable the change consumer on the enable toggle?

**Comment by gtmacdonald** [comment]
@tardyp I've finished the change requests except for your question about stopping the scheduler. I'll do some more testing to see if the change requests are backing up.

**Comment by gtmacdonald** [comment]
@tardyp I think I understand now what you're asking for about stopping the scheduler. The current implementation suits my purposes with the timed scheduler. Please let me know if you'd like me to change it.

**Comment by tardyp** [comment]
@gtmacdonald I think indeed, it makes much more sense to just temporarily stop the scheduler so that it can keep up later.

For me another use case for the scheduler disabling is if you have some issue with the infra and do not want more spurious failed builds while repairing it (that might also be your usecase, I am not sure)
When you enable it back you would expect to have buildbot keep up with the changes that were recorded in between. For TimedScheduler, this will be equivalent as your current implementation

Usually people would just stop their master, but then this also stop the change source, and you will miss builds.

**Comment by gtmacdonald** [comment]
That makes sense. In our case we sometimes need to shut off a long running build to switch to a quicker incremental one; otherwise, for legacy reasons, they'll end up interfering.

So will I need to add a queue for the missed changes in the scheduler? Or is there a method call to stop the scheduler but keep it registered for changes?

**Comment by tardyp** [comment]
You dont need to do anything, the scheduler is already keeping the id of the last change it processed. At restart it will start from there. I think you can just call deactivate, and in activate make sure the scheduler is not disabled.

**Comment by gtmacdonald** [comment]
I think it's good 2 go. Please let me know if there's anything else.

**Comment by tardyp** [comment]
this needs some rebase :-}

File: www/base/src/app/schedulers/schedulers.controller.coffee [hunk 17]
```diff
@@ -2,3 +2,10 @@ class schedulers extends Controller
     constructor: ($log, $scope, $location, dataService) ->
         data = dataService.open().closeOnDestroy($scope)
         $scope.schedulers = data.getSchedulers()
+
+        $scope.change = (s) ->
+        	newValue = s.enabled
+        	param = enabled: newValue
+        	res = dataService.control 'schedulers', s.schedulerid, 'enable', param
+        	res.then (-> $log.debug("succeeded")), (data) => $log.error(data.error)
```
Comment (tardyp) [inline www/base/src/app/schedulers/schedulers.controller.coffee]: no need for logging

**Comment by gtmacdonald** [comment]
how's it look now? BTW, do I need to comment here to alert you that it's ready? Or do you get an automated message?

**Comment by gtmacdonald** [comment]
This should be g2g now. :)

File: master/buildbot/test/unit/test_schedulers_base.py [hunk 18]
```diff
@@ -158,9 +158,10 @@ def gotChange(got_change, got_important):
...
 
         def test(_):
-            # check that it registered a callback
-            self.assertEqual(len(self.mq.qrefs), 1)
-            qref = self.mq.qrefs[0]
+            # check that it registered callbacks
+            self.assertEqual(len(self.mq.qrefs), 2)
+
+            qref = self.mq.qrefs[1]
             self.assertEqual(qref.filter, ('changes', None, 'new'))
```
Comment (tardyp) [inline master/buildbot/test/unit/test_schedulers_base.py]: There is a need to test that the scheduler correctly react to the disable events

File: master/buildbot/schedulers/trysched.py [hunk 19]
```diff
@@ -117,6 +120,9 @@ def activate(self):
 
     @defer.inlineCallbacks
     def deactivate(self):
+        if not self.enabled:
```
Comment (tardyp) [inline master/buildbot/schedulers/trysched.py]: base scheduler calls this method when a disable message is received, so the disable message will not actually disable.

Actually, this might work because self.enabled is set after calling the method.
..but the calling does not wait for results, and inlineCallback does not guarantee that the decorated method is called synchronously.

A unit test should test that to make sure it works for every version of twisted/python.

File: master/buildbot/schedulers/base.py [hunk 20]
```diff
@@ -90,15 +90,38 @@ def __init__(self, name, builderNames, properties=None,
...
+
+        if not self._enable_consumer:
+            yield self.startConsumingEnableEvents()
+
+    def _enabledCallback(self, key, msg):
+        if msg['enabled']:
+            self.enabled = True
+            self.activate()
+        else:
+            self.deactivate()
```
Comment (tardyp) [inline master/buildbot/schedulers/base.py]: should wait for result of deactivate before actually disable

File: master/buildbot/schedulers/trysched.py (hunk 19 above)
Comment (gtmacdonald) [inline master/buildbot/schedulers/trysched.py]: I think waiting for the deactivate as you mention below should fix this. I just don't want the enabled flag to be out of sync with reality.

File: master/buildbot/test/unit/test_schedulers_base.py (hunk 18 above)
Comment (gtmacdonald) [inline master/buildbot/test/unit/test_schedulers_base.py]: So if I defer the enabled state like we're planning, I'm not sure how to wait for the enabled state to change in the test. This is what I was planning on doing to defer the enabled state:

    def _enabledCallback(self, key, msg):
        if msg['enabled']:
            self.enabled = True
            self.activate()
        else:
            d = self.deactivate()
            def fn():
                self.enabled = False
            d.addCallback(fn)

Any suggestions? I'm not quite comfortable with this reactor pattern yet.

Comment (tardyp) [inline master/buildbot/test/unit/test_schedulers_base.py]: You can call the callback directly from the tests instead of using the mq api to send the message. you need to make your test inlineCallbacks and then wait for the results of the callback in the test and make the appropriate assertion after that.
There are plenty of examples in the current tests

Comment (gtmacdonald) [inline master/buildbot/test/unit/test_schedulers_base.py]: gotcha, thanks!

File: master/buildbot/schedulers/trysched.py (hunk 19 above)
Comment (gtmacdonald) [inline master/buildbot/schedulers/trysched.py]: How should I go about testing each version of twisted/python? Is there an example of this somewhere?

Comment (tardyp) [inline master/buildbot/schedulers/trysched.py]: We have a CI so that your unit tests are run against all version of twisted/python.

So you just have to write the unit tests, and then they will run everywhere when you push your branch

Comment (gtmacdonald) [inline master/buildbot/schedulers/trysched.py]: oh that's great!

**Comment by gtmacdonald** [comment]
So I think I've added the last unit test. One of the checks isn't passing but it seems unrelated.

**Comment by tardyp** [comment]
@gtmacdonald please install codecov.io chrome extension, or use the website to checkout the coverage. There are still too many corner case which are not covered by the unit tests.

**Comment by gtmacdonald** [comment]
@tardyp Hi Pierre, When I try to see the coverage using the chrome extension I get an error. It says something like, "there was a problem with the coverage, sorry." I tried running it on my own branch but I can't figure out how to generate the .coverage files.

**Comment by gtmacdonald** [comment]
Doing a fresh rebase and push to see if I can get the code coverage tool to work.

**Comment by gtmacdonald** [comment]
I like that codecov tool.

Should be g2g now. :)

**Comment by tardyp** [comment]
One of the longest to land PR. congrats! thanks for keeping up!

**Comment by gtmacdonald** [comment]
lol, thanks!

**Comment by tardyp** [comment]
@gtmacdonald I got a report from use for a security problem with this PR.

Any anonymous user can disable a scheduler with this.
We need to add an endpointmatcher for this action, and make sure the doc is updated to explain how to require administrator login to disable schedulers

**Comment by gtmacdonald** [comment]
Ok, I'll start looking into it.

**Comment by tardyp** [comment]
@gtmacdonald I did implement this https://github.com/buildbot/buildbot/pull/2822

I would like you to test it with your environment, and make sure it works beyond the unit tests

**Comment by gtmacdonald** [comment]
Sure thing. Thanks for implementing it. :)

**Comment by gtmacdonald** [comment]
This looks like it works well with one root user and the following config

c['www'] = dict(port=8010,                                                                             
                plugins=dict(waterfall_view={}, console_view={}))                                      
                                                                                                       
authz = util.Authz(                                                                                    
    allowRules=[util.AnyEndpointMatcher(role="admins")],                                               
    roleMatchers=[util.RolesFromUsername(roles=['admins'], usernames=['root'])]                        
)                                                                                                      
                                                                                                       
auth=util.UserPasswordAuth({'root': 'pass'})                                                           
                                                                                                       
c['www']['auth'] = auth                                                                                
c['www']['authz'] = authz 

If i switch AnyEndpointMatcher to EnabledSchedulerEndpointMatcher. It still works but when I log out and switch back to anonymous, the UI is still there. The check boxes are ineffectual until I log back in as root though. Aside from this UI glitch it seems to be working.

**Comment by gtmacdonald** [comment]
I'll try with a regular non-admin user shortly.

**Comment by gtmacdonald** [comment]
Ok, if I add another user instead of just root everything works as expected. The UI clears immediately after I logout.

**Comment by gtmacdonald** [comment]
so this fixed it:

auth=util.UserPasswordAuth({'root': 'pass', 'greg': 'pass'})

**Comment by gtmacdonald** [comment]
It'd be nice if there was authentication that plugged into Microsoft's user management stuff.

**Comment by gtmacdonald** [comment]
Idk, do you think we can ship it as is or should we fix this bug. It's kind of small. Might confuse build admins though.

**Comment by tardyp** [comment]
If you can come up with a fix on the UI which displays an error message if
user is not allowed to change the scheduler (and put the state of the
checkbox back), that'd be nice.


Le lun. 6 mars 2017 à 22:25, Greg MacDonald <notifications@github.com> a
écrit :

**Comment by gtmacdonald** [comment]
I think the proper fix is for it to remove those UI elements after it logs out of root. Like it does when there's more than just the root user. So the logic is there it's just not happening on an edge case for some reason. I'll "root" around for it tomorrow. :D ha ha i'm so funny.
